mava/systems/sable/sebulba/rec_sable.py

```diff
@@ -0,0 +1,787 @@
+        config (DictConfig): Configuration settings for the environment and rollout.
+        rollout_queue (Pipeline): Queue for sending collected rollouts to the learner.
+        params_source (ParamsSource): Source for fetching the latest network parameters
+        from the learner.
+        apply_fns (Tuple): Functions for running the actor and critic networks.
+        actor_device (Device): Actor device to use for rollout.
+        seeds (List[int]): Seeds for initializing the environment.
+        stop_event (threading.Event): Manages the thread's lifecycle.
+    """
+    name = threading.current_thread().name
+    print(f"{Fore.BLUE}{Style.BRIGHT}Thread {name} started{Style.RESET_ALL}")
+
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def act_fn(
+        params: Params,
+        observation: Observation,
+        hstates: chex.Array,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+        action, log_prob, value, hstates = apply_fns(  # type: ignore
+            params,
+            observation=observation,
+            hstates=hstates,
+            key=key,
+        )
+        return action, log_prob, value, hstates
+
+    timestep = env.reset(seed=seeds)
+
+    hstates = get_init_hidden_state(config.network.net_config, config.arch.num_envs)
+
+    # Loop till the desired num_updates is reached.
+    while not stop_event.is_set():
+        # Rollout
+        traj: List[Transition] = []
+        episode_metrics: List[Dict] = []
+        actor_timings: Dict[str, List[float]] = defaultdict(list)
+
+        # Store the first hstates
+        prev_hstates = hstates
+
+        with RecordTimeTo(actor_timings["rollout_time"]):
+            for _ in range(config.system.rollout_length):
+                with RecordTimeTo(actor_timings["get_params_time"]):
+                    params = params_source.get()  # Get the latest parameters from the learner
+
+                obs_tpu = tree.map(move_to_device, timestep.observation)
+
+                prev_dones = jnp.repeat(timestep.last(), config.system.num_agents).reshape(
+                    config.arch.num_envs, -1
+                )
+
+                # Get action and value
+                with RecordTimeTo(actor_timings["compute_action_time"]):
+                    key, act_key = jax.random.split(key)
+                    action, log_prob, value, hstates = act_fn(params, obs_tpu, hstates, act_key)
+                    cpu_action = jax.device_get(action)
+
+                # Step environment
+                with RecordTimeTo(actor_timings["env_step_time"]):
+                    timestep = env.step(cpu_action)
+
+                # Updated the dones and Hstates
+                dones = timestep.last()
+                dones = jnp.expand_dims(dones, (1, 2, 3, 4))
+                hstates = tree.map(
+                    lambda hs, dones=dones: jnp.where(dones, jnp.zeros_like(hs), hs), hstates
+                )
```

Comment:
np instead of jnp

```suggestion
                dones = np.expand_dims(dones, (1, 2, 3, 4))
                hstates = tree.map(
                    lambda hs, dones=dones: np.where(dones, np.zeros_like(hs), hs), hstates
                )
```